x-pack/plugins/lists/server/services/extension_points/extension_point_storage.ts

```diff
@@ -0,0 +1,157 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+/* eslint-disable max-classes-per-file */
+
+import type { Logger } from 'kibana/server';
+
+import { ExtensionPoint } from './types';
+import { ExtensionPointError } from './errors';
+
+type NarrowExtensionPointToType<T extends ExtensionPoint['type']> = { type: T } & ExtensionPoint;
+
+export class ExtensionPointStorage {
+  private readonly store = new Map<ExtensionPoint['type'], Set<ExtensionPoint>>();
+  private readonly registeredFrom = new Map<ExtensionPoint, string>();
+
+  constructor(private readonly logger?: Logger) {}
+
+  add(extension: ExtensionPoint): void {
+    if (!this.store.has(extension.type)) {
+      this.store.set(extension.type, new Set());
+    }
+
+    const extensionPointsForType = this.store.get(extension.type);
+
+    if (extensionPointsForType) {
+      extensionPointsForType.add(extension);
+
+      // Capture stack trace from where this extension point was registered, so that it can be used when
+      // errors occur or callbacks don't return the expected result
+      const from = new Error('REGISTERED FROM:').stack ?? 'REGISTERED FROM: unknown';
```

Comment:
I think it will be important to understand where exceptions are coming from when calling external callbacks. This here captures the `stack` when registering the extension point and is used further below when running the callbacks if they `throw` or if the data they return is not valid.

<details><summary>CLICK HERE TO SEE EXAMPLE OF LOG MESSAGE</summary>

```
[2021-12-21T11:30:58.992-05:00][ERROR][plugins.lists] ExtensionPointError: Extension point for exceptionsListPreCreateItem returned data that failed validation: REGISTERED FROM:
    at ExtensionPointStorage.add (/Users/ptavares/WORKSPACES/kibana/x-pack/plugins/lists/server/services/extension_points/extension_point_storage.ts:35:20)
    at registerListsServerExtension (/Users/ptavares/WORKSPACES/kibana/x-pack/plugins/lists/server/plugin.ts:82:30)
    at EndpointAppContextService.start (/Users/ptavares/WORKSPACES/kibana/x-pack/plugins/security_solution/server/endpoint/endpoint_app_context_services.ts:127:7)
    at Plugin.start (/Users/ptavares/WORKSPACES/kibana/x-pack/plugins/security_solution/server/plugin.ts:414:36)
    at PluginWrapper.start (/Users/ptavares/WORKSPACES/kibana/src/core/server/plugins/plugin.ts:127:41)
    at PluginsSystem.startPlugins (/Users/ptavares/WORKSPACES/kibana/src/core/server/plugins/plugins_system.ts:186:40)
    at PluginsService.start (/Users/ptavares/WORKSPACES/kibana/src/core/server/plugins/plugins_service.ts:199:56)
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at Server.start (/Users/ptavares/WORKSPACES/kibana/src/core/server/server.ts:331:5)
    at Root.start (/Users/ptavares/WORKSPACES/kibana/src/core/server/root/index.ts:62:14)
    at bootstrap (/Users/ptavares/WORKSPACES/kibana/src/core/server/bootstrap.ts:102:5)
    at Command.<anonymous> (/Users/ptavares/WORKSPACES/kibana/src/cli/serve/serve.js:244:5)
    ----------------------
    at ExtensionPointStorageClient.pipeRun (/Users/ptavares/WORKSPACES/kibana/x-pack/plugins/lists/server/services/extension_points/extension_point_storage.ts:134:13)
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at ExceptionListClient.createExceptionListItem (/Users/ptavares/WORKSPACES/kibana/x-pack/plugins/lists/server/services/exception_lists/exception_list_client.ts:432:18)
    at /Users/ptavares/WORKSPACES/kibana/x-pack/plugins/lists/server/routes/create_exception_list_item_route.ts:91:33
    at Router.handle (/Users/ptavares/WORKSPACES/kibana/src/core/server/http/router/router.ts:275:30)
    at handler (/Users/ptavares/WORKSPACES/kibana/src/core/server/http/router/router.ts:230:13)
    at exports.Manager.execute (/Users/ptavares/WORKSPACES/kibana/node_modules/@hapi/hapi/lib/toolkit.js:60:28)
    at Object.internals.handler (/Users/ptavares/WORKSPACES/kibana/node_modules/@hapi/hapi/lib/handler.js:46:20)
    at exports.execute (/Users/ptavares/WORKSPACES/kibana/node_modules/@hapi/hapi/lib/handler.js:31:20)
    at Request._lifecycle (/Users/ptavares/WORKSPACES/kibana/node_modules/@hapi/hapi/lib/request.js:371:32)
    at Request._execute (/Users/ptavares/WORKSPACES/kibana/node_modules/@hapi/hapi/lib/request.js:281:9)

```

</details>